File: EstruturaDeDados/Lista/DoubleLinkedList/Java/ListaDuplamenteEncadeada.java
public class ListaDuplamenteEncadeada{

    private class Nodo{
        private int valor;
        private Nodo prox;
        private Nodo prev;
        public Nodo(int valor) {
            this.valor = valor;
            this.prox = null;
            this.prev = null;
        }
        public int getValor() {
            return valor;
        }
        public void setValor(int valor) {
            this.valor = valor;
        }
        public Nodo getProx() {
            return this.prox;
        }
        public void setProx(Nodo referencia) {
            this.prox = referencia;
        }        
        public Nodo getPrev() {
            return this.prox;
        }
        public void setPrev(Nodo referencia) {
            this.prev = referencia;
        }        
    }

    private Nodo inicio, fim;
    private int size;

    public ListaDuplamenteEncadeada(){
        clear();
    }

    public boolean add(int value){
        Nodo novoObjeto = new Nodo(value);

        if(isEmpty()){
            inicio=novoObjeto;            
        }
        else{
            fim.setProx(novoObjeto);
            novoObjeto.setPrev(fim);
        }
        fim=novoObjeto;
        size++;
        return true;
    }

    // remove o último elemento da lista
    public boolean remove(){
    }

    public int size(){
        return size;
    }

    public boolean isEmpty(){
        return (this.size==0);
    }

    public int getDoInicio(int idx){
        if( (idx>=0) && (idx<size) ){
            Nodo amyrKlink = inicio;
            while(idx>0){
                amyrKlink=amyrKlink.getProx();
                idx--;
            }
            return amyrKlink.getValor();
        }
        return -1;
    }

    public int getDoFim(int idx){swapNodo
            while(novoIdx>0){
                amyrKlink=amyrKlink.getPrev();
                novoIdx--;
            }
            return amyrKlink.getValor();

        }
        return -1;
    }

    public void clear(){
        inicio=fim=null;
        size=0;
    }

    public boolean addAt(int idx, int value){
        if((idx>=0) && (idx<=size)){

            if(idx==0){
                if(isEmpty())
                  add(value);
                else{
                    Nodo novoNodo = new Nodo(value);
                    inicio.setPrev(novoNodo);
                    novoNodo.setProx(inicio);
                    inicio=novoNodo;
                    size++;
                }
            }
            else if(idx==size()){
                add(value);
            }
            else{
                Nodo amyrKlink = inicio;
                while(idx>0){
                    amyrKlink=amyrKlink.getProx();
                    idx--;
                }
                Nodo novoObjeto = new Nodo(value);
                novoObjeto.setProx(amyrKlink);
                novoObjeto.setPrev(amyrKlink.getPrev());
                amyrKlink.setPrev(novoObjeto);
                novoObjeto.getPrev().setProx(novoObjeto);

                size++;
            }
            

            return true;
        }
        return false;
    }

    public boolean removeAt(int idx){        
    }

    public boolean swapValor(int idx01, int idx02){
    }

    public boolean swapNodo(int idx01, int idx02){

        if(
            (idx01>=0 && idx01<size) && 
            (idx02>=0 && idx02<size) &&
            (idx01 != idx02)
          )
          {
            // swap de nodos no meio (nem fim nem início)

            Nodo A, B;
            A=inicio;
            while(idx01>0){
                A=A.getProx();
                idx01--;
            }

            B=inicio;
            while(idx02>0){
                B=B.getProx();
                idx02--;
            }

            //swap de fato
            if(A!=fim)
                A.getProx().setPrev(B);
            if(B!=fim)
                B.getProx().setPrev(A);

            if(A!=inicio)
                A.getPrev().setProx(B);
            if(B!=inicio)
                B.getPrev().setProx(A);

            Nodo auxiliar = A.getPrev();
            A.setPrev(B.getPrev());
            B.setPrev(auxiliar);

            auxiliar= A.getProx();
            A.setProx(B.getProx());
            B.setProx(auxiliar);

            // Ajustes do inicio e fim da lista
            if(A==inicio || A==fim){                
                if(B==inicio || B==fim){
                    if(A==inicio && B==fim){
                        inicio=B;
                        fim=A;
                    }
                    else{
                        inicio=A;
                        fim=B;
                    }
                }
                else{
                    if(A==inicio)
                        inicio=B;
                    else
                        fim=B;
                }
            }
            else if(B==inicio || B==fim){
                if(B==inicio)
                    inicio=A;
                else
                    fim=A;
    }




          }

    }

}

File: EstruturaDeDados/Lista/DoubleLinkedList/cpp/ListaDuplamenteEncadeada.cpp
#include "ListaDuplamenteEncadeada.hpp"

ListaDuplamenteEncadeada::ListaDuplamenteEncadeada(){
    nElementos=0;
    fim=inicio=nullptr;
};

bool ListaDuplamenteEncadeada::add(int value){

    if(isEmpty()){
        fim = inicio = new Nodo(value);
    }
    else{
        Nodo * novoNodo = new Nodo(value);
        novoNodo->setAnterior(fim);
        fim->setProximo(novoNodo);
        fim=novoNodo;
    }
    nElementos++;
}

bool ListaDuplamenteEncadeada::remove(int pos){
    if(pos>=0 && pos<nElementos){
        //busca do nodo a ser eliminado
        Nodo * buscaNodo = inicio;
        while(pos<0){
            buscaNodo=buscaNodo->getProximo();
            pos--;
        }

        if(buscaNodo!=fim)
            buscaNodo->getProximo()->setAnterior(buscaNodo->getAnterior());

        if(buscaNodo!=inicio)
            buscaNodo->getAnterior()->setProximo(buscaNodo->getProximo());

        if(buscaNodo==inicio)
            inicio=buscaNodo->getProximo();
        if(buscaNodo==fim)
            fim=buscaNodo->getAnterior();

        buscaNodo->setAnterior(nullptr);
        buscaNodo->setProximo(nullptr);
        delete buscaNodo;

        nElementos--;

        return true;

    }
    else
        return false;       

}

bool ListaDuplamenteEncadeada::swap(int idx01, int idx02){

    if(
        // indice 1 é valido
        (idx01>=0 && idx01<nElementos) &&
        // indice 2 é valido
        (idx02>=0 && idx02<nElementos) &&
        // indice 1 é diferente de indice 2
        (idx01!=idx02)
    ){
        //busca o primeiro nodo
        Nodo * A=inicio;
        while(idx01>0){
            A=A->getProximo();
            idx01--;
        }

        //busca o segundo nodo
        Nodo * B=inicio;
        while(idx02>0){
            B=B->getProximo();
            idx02--;
        }

        if(A->getProximo()!=nullptr)
            A->getProximo()->setAnterior(B);
        if(B->getProximo()!=nullptr)
            B->getProximo()->setAnterior(A);
        if(A->getAnterior()!=nullptr)
            A->getAnterior()->setProximo(B);
        if(B->getAnterior()!=nullptr)
            B->getAnterior()->setProximo(A);

        Nodo * aux;
        aux=A->getProximo();
        A->setProximo(B->getProximo());
        B->setProximo(aux);

        aux=A->getAnterior();
        A->setAnterior(B->getAnterior());
        B->setAnterior(aux); 

        if(A==inicio || A==fim){
            if(B==inicio || B==fim){
                if(A==fim){
                    fim=B;
                    inicio=A;
                }
                else{
                    fim=A;
                    inicio=B
                }
            }
            else{
                if(A==fim)
                    fim=B;
                else
                    inicio=B;
            }
        }
        else if(B==inicio || B==fim){
            if(B==fim)
                fim=A;
            else
                inicio=A;
        }


        return true;
    }
    else
        return false;

}

int ListaDuplamenteEncadeada::size(){
}

bool ListaDuplamenteEncadeada::isEmpty(){
    return (nElementos==0);
}

int ListaDuplamenteEncadeada::get(int idx){
}

void ListaDuplamenteEncadeada::clear(){
}
// ==> métodos do nodo base da lista
ListaDuplamenteEncadeada::Nodo::Nodo(int valor){
    this->valor=valor;
    this->proximo=this->anterior=nullptr;
}
int ListaDuplamenteEncadeada::Nodo::getValor(){
    return this->valor;
}
void ListaDuplamenteEncadeada::Nodo::setValor(int valor){
    this->valor = valor;    
}
ListaDuplamenteEncadeada::Nodo * ListaDuplamenteEncadeada::Nodo::getProximo(){
    return this->proximo;
}
void ListaDuplamenteEncadeada::Nodo::setProximo(Nodo* referencia){
    this->proximo = referencia;
}
ListaDuplamenteEncadeada::Nodo * ListaDuplamenteEncadeada::Nodo::getAnterior(){
    return this->anterior;
}
void ListaDuplamenteEncadeada::Nodo::setAnterior(Nodo* referencia){
    this->anterior = referencia;
}
